core/src/com/unciv/logic/automation/unit/CivilianUnitAutomation.kt

```diff
@@ -117,6 +123,33 @@ object CivilianUnitAutomation {
                 UnitActions.invokeUnitAction(unit, UnitActionType.StartGoldenAge)
         }
 
+        if (unit.hasUnique(UniqueType.GainFreeBuildings)) {
+            val unique = unit.getMatchingUniques(UniqueType.GainFreeBuildings).first()
+            val buildingName = unique.params[0]
+            // Choose the city that is closest in distance and does not have the building constructed.
+            val cityToGainBuilding = unit.civ.cities.filter {
+                !it.cityConstructions.containsBuildingOrEquivalent(buildingName)
+                    && (unit.movement.canMoveTo(it.getCenterTile()) || unit.currentTile == it.getCenterTile())
+            }.minByOrNull {
+                val path = unit.movement.getShortestPath(it.getCenterTile())
+                path.size
+            }
+
+            if (cityToGainBuilding != null) {
```

Comment:
To avoid mid modders trying to take advantage of this, I feel we should maybe make use of the building rank function I added in ConstructionAutomation. If that function needs to be moved to Automation, that's fine

Reply:
> I feel we should maybe make use of the building rank function I added in ConstructionAutomation. 

Where is the building rank function you mentioned 👀 ? I only saw `chooseNextConstruction` function

Reply:
You'll have to merge the current master into this PR. Or, actually, this isn't necessarily blocking, so you can work on that as a separate PR